internal/page/page.go

```diff
@@ -0,0 +1,111 @@
+package page
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+
+	"github.com/RHEnVision/provisioning-backend/internal/math"
+
+	"github.com/RHEnVision/provisioning-backend/internal/ptr"
+	"github.com/rs/zerolog"
+)
+
+type (
+	ctxKeyType  int
+	limitOffset int
+)
+
+const (
+	offsetCtxKey ctxKeyType = iota
+	limitCtxKey
+)
+
+const (
+	defaultValueLimit  = 100
+	defaultValueOffset = 0
+)
+
+// WithOffset returns context copy with offset value.
+func WithOffset(ctx context.Context, offsetStr string) context.Context {
+	logger := zerolog.Ctx(ctx)
+	offset, err := strconv.Atoi(offsetStr)
+	if err != nil || offset < 0 {
+		logger.Err(err).Msg("Offset is missing or invalid, setting offset value to 0")
+		offset = defaultValueOffset
+	}
+	return context.WithValue(ctx, offsetCtxKey, offset)
+}
+
+// WithLimit returns context copy with limit value.
+func WithLimit(ctx context.Context, limitStr string) context.Context {
+	logger := zerolog.Ctx(ctx)
+	limit, err := strconv.Atoi(limitStr)
+	if err != nil || limit < 0 {
+		logger.Err(err).Msg("Limit is missing or invalid, setting limit value to 100")
+		limit = defaultValueLimit
+	}
+	return context.WithValue(ctx, limitCtxKey, limit)
+}
+
+func Limit(ctx context.Context) limitOffset {
+	if lim, ok := ctx.Value(limitCtxKey).(int); ok {
+		return limitOffset(lim)
+	}
+	return limitOffset(defaultValueLimit)
+}
+
+func Offset(ctx context.Context) limitOffset {
+	if off, ok := ctx.Value(offsetCtxKey).(int); ok {
+		return limitOffset(off)
+	}
+	return limitOffset(defaultValueOffset)
+}
+
+func (o limitOffset) IntPtr() *int {
+	return ptr.To(int(o))
+}
+
+func (o limitOffset) Int() int {
+	return int(o)
+}
+
+func (o limitOffset) Int64() int64 {
+	return int64(o)
+}
+
+func (o limitOffset) String() string {
+	return strconv.Itoa(int(o))
+}
+
+type Metadata struct {
+	Previous string `json:"previous,omitempty"`
+	Next     string `json:"next,omitempty"`
+}
+
+func MetadataResponse(ctx context.Context, r *http.Request, total int) *Metadata {
+	limit := Limit(ctx).Int()
+	offset := Offset(ctx).Int()
+	nextOffset := math.Min(total, offset+limit)
+	prevOffset := math.Max(0, offset-limit)
+
+	q := r.URL.Query()
+
+	// set the "prev" link with offset - limit or 0
+	q.Set("limit", strconv.Itoa(limit))
+	q.Set("offset", strconv.Itoa(prevOffset))
+	params, _ := url.PathUnescape(q.Encode())
+	prev := fmt.Sprintf("%v?%v", r.URL.Path, params)
+
+	// set the "next" link with limit+offset or total set for the next page
+	q.Set("offset", strconv.Itoa(nextOffset))
+	params, _ = url.PathUnescape(q.Encode())
+	next := fmt.Sprintf("%v?%v", r.URL.Path, params)
```

Comment:
Uh, this is very ugly. Let’s take a look how you can cleanly create URL parameters in Go: 

```
	q := url.Values{}
	q.Add("clean", "abc")
	q.Add("dirty", "100% $999")
	q.Add("base64", "123/ABC+abc=")

	// prints: base64=123%2FABC%2Babc%3D&clean=abc&dirty=100%25+%24999
	println(q.Encode())
```

Here is a working example to play with: https://go.dev/play/p/eQU_2AZoI5W

Now, the main issue with base64 encoding is that it contains three characters which have special meaning in URLs: `/+=` and these needs to be escaped. There is no need to ever use PathUnescape, I am assuming that the AWS client library does perform proper URL decoding and return the token in the correct (normal) format. Something like `123/ABC+abc=` and not `123%2FABC%2Babc%3D` therefore you can just simply build a new params and call `Encode` to get the proper result.

My suggestion: just use `url.Values` and `Add`/`Encode` methods and it will work fine.